### PR TITLE
Add "prepare" and "check" watchers

### PR DESCRIPTION
**Pull request by mergeconflict**

### CODE REVIEW

File: evwatcher.c
Comment (htuch) [inline evwatcher.c]: Nit: I do see plenty of `return 0;` style code in `event.c`, I'd prefer to not use these superfluous parens if we can skip.

File: evwatcher.c
Comment (htuch) [inline evwatcher.c]: Ditto etc.

File: include/event2/event_watcher.h
Comment (htuch) [inline include/event2/event_watcher.h]: I'd let the work on epoll delay/duration drive this. If libuv keeps it simple, seems fine to go that way.

File: include/event2/event_watcher.h
Comment (htuch) [inline include/event2/event_watcher.h]: Do we also need to update http://www.wangafu.net/~nickm/libevent-book/ with this? That's basically the libevent docs for most folks. What about release history version files?

File: evwatcher.c
Comment (htuch) [inline evwatcher.c]: Should we be using the `mm_alloc` family of functions here? What about possibly `NULL` return? See for example `evbuffer_chain_new` and its handling.

File: test/test-prepare-check.c
Comment (htuch) [inline test/test-prepare-check.c]: Nit: weird indent. In fact, the indent in the the test file is generally shorter than the implementation..

File: test/test-prepare-check.c
Comment (htuch) [inline test/test-prepare-check.c:152]: More weird indent..

**Review by htuch — CHANGES_REQUESTED** [review]
This looks great overall, ready to ship to libevent GH with some minor comments.

File: event.c
Comment (mergeconflict) [inline event.c]: This is definitely in the wrong place. It should be further down, after `evsel->dispatch(...)`

File: event.c
Comment (mergeconflict) [inline event.c]: Not sure if this should be here, or just before `evsel->dispatch(...)`. TBD.

File: include/event2/event_watcher.h
Comment (mergeconflict) [reply include/event2/event_watcher.h]: I've added a new `event_watcher_cb_info` struct containing timing information. I'm pretty sure it's actually necessary; there isn't any other good way of determining the timeout value computed by `timeout_next` in `event_base_loop` from outside.

File: include/event2/event_watcher.h
Comment (mergeconflict) [reply include/event2/event_watcher.h]: I'm not sure about the book - it's not in the same repository. And there is a changelog, but it seems to have bitrotted... the last entry is from January 2017. We can get guidance from the libevent maintainers.

File: event.c
Comment (htuch) [inline event.c]: Do you want separate CB info for prepare/check?

File: test/test-prepare-check.c
Comment (htuch) [inline test/test-prepare-check.c:128]: These tests catch the regression you spotted around check/dispatch ordering, I'm guessing, can you call that out in a comment somewhere?

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo comments.

File: event.c
Comment (mergeconflict) [reply event.c]: I didn't see an immediate need for that, but on second thought, the whole purpose of struct-ifying these is for future flexibility... So yes, yes I want that. Thanks!

File: test/test-prepare-check.c
Comment (mergeconflict) [reply test/test-prepare-check.c:128]: Yeah, I should write a summary at the top... there's a fair bit going on here.

**Pull request closed without merging**
